modules/data.remote/inst/satellitetools/biophys_xarray.py

```diff
@@ -0,0 +1,408 @@
+
+    """
+    b5 = dataset.band_data.sel(band="B5")
+    b7 = dataset.band_data.sel(band="B7")
+    ci_red_edge = (b7 / b5) - 1
+    return dataset.assign({"ci_red_edge": ci_red_edge})
+
+
+def run_snap_biophys(dataset, variable):
+    """Compute specified variable using the SNAP algorithm.
+
+    See ATBD at https://step.esa.int/docs/extra/ATBD_S2ToolBox_L2B_V1.1.pdf
+
+    Parameters
+    ----------
+    dataset : xr dataset
+        xarray dataset.
+    variable : str
+        Options 'FAPAR', 'FCOVER', 'LAI', 'LAI_Cab' or 'LAI_Cw'
+
+    Returns
+    -------
+    xarray dataset
+        Adds the specified variable array to dataset (variable name in
+        lowercase).
+
+    """
+    # generate view angle bands/layers
+    vz = (
+        np.ones_like(dataset.band_data[:, 0, :, :]).T
+        * np.cos(np.radians(dataset.view_zenith)).values
+    )
+    vz = vz[..., np.newaxis]
+    vzarr = xr.DataArray(
+        vz,
+        coords=[dataset.y, dataset.x, dataset.time, ["view_zenith"]],
+        dims=["y", "x", "time", "band"],
+    )
+
+    sz = (
+        np.ones_like(dataset.band_data[:, 0, :, :]).T
+        * np.cos(np.radians(dataset.sun_zenith)).values
+    )
+    sz = sz[..., np.newaxis]
+    szarr = xr.DataArray(
+        sz,
+        coords=[dataset.y, dataset.x, dataset.time, ["sun_zenith"]],
+        dims=["y", "x", "time", "band"],
+    )
+
+    raz = (
+        np.ones_like(dataset.band_data[:, 0, :, :]).T
+        * np.cos(np.radians(dataset.sun_azimuth - dataset.view_azimuth)).values
+    )
+    raz = raz[..., np.newaxis]
+    razarr = xr.DataArray(
+        raz,
+        coords=[dataset.y, dataset.x, dataset.time, ["relative_azimuth"]],
+        dims=["y", "x", "time", "band"],
+    )
+
+    newarr = xr.concat([dataset.band_data, vzarr, szarr, razarr], dim="band")
+    newarr = newarr.stack(xy=("x", "y"))
+    arr = xr.apply_ufunc(
+        _compute_variable,
+        newarr,
+        input_core_dims=[["band", "xy"]],
+        output_core_dims=[["xy"]],
+        kwargs={"variable": variable},
+        vectorize=True,
+    ).unstack()
+    return dataset.assign({variable.lower(): arr})
+
+
+def estimate_gpp_vi_lue(vi, daily_par, model_name="ci_red_edge_1"):
+    """Estimate GPP using simple vegetation index based models and PAR.
+
+    This function has not been properly tested (i.e.used for a while)
+
+    Parameters
+    ----------
+    vi : float
+        Vegetation index values.
+    daily_par : float
+        Daily PAR as MJ/s/m².
+    model_name : str, optional
+        Name of the model (see biophys_xarray.GPP_LUE_models).
+        The default is 'ci_red_edge_1'.
+
+    Returns
+    -------
+    gpp : float
+        Estimated gross primary productivity.
+
+    """
+    vi_name = "_".join(model_name.split("_")[:-1])
+    gpp = GPP_LUE_models[vi_name][model_name]["model"](
+        np.array(vi), np.array(daily_par)
+    )
+    return gpp
+
+
+# GPP estimation models
+GPP_LUE_models = {
```

Comment:
This dictionary containing the model details is not required.

Reply:
let's remove